extension/inet/inet_escape_functions.cpp

```diff
@@ -0,0 +1,491 @@
+			auto result_string = StringVector::EmptyString(result, total_chars_in_list);
+			auto result_data = result_string.GetDataWriteable();
+			Match *current_match = head;
+			while (current_match != nullptr) {
+				D_ASSERT(current_match->start - next_match_pos > 0);
+				// copy the characters infront or between the matches
+				memcpy(result_data + result_pos, input_data + next_match_pos,
+				       current_match->start - 1 - next_match_pos);
+				result_pos += current_match->start - 1 - next_match_pos;
+				next_match_pos = current_match->start + current_match->size + current_match->semicolon;
+
+				// copy the decoded result
+				memcpy(result_data + result_pos, current_match->decoded.GetData(), current_match->decoded.GetSize());
+				result_pos += current_match->decoded.GetSize();
+				// next match
+				current_match = current_match->next;
+			}
+			memcpy(result_data + result_pos, input_data + next_match_pos, input_size - next_match_pos);
+			result_string.Finalize();
+			return result_string;
+		}
+		auto result_string = StringVector::EmptyString(result, 0);
+		result_string.Finalize();
+		return result_string;
+	}
+
+private:
+	void append(idx_t start, idx_t size, bool semi) {
+		D_ASSERT(start + size + semi <= input.GetSize());
+		auto new_match = create_match(start, size, semi);
+		if (head == nullptr) {
+			head = new_match;
+		} else {
+			Match *current = head;
+			while (current->next != nullptr) {
+				current = current->next;
+			}
+			current->next = new_match;
+		}
+		total_chars_in_list += (new_match->start - 1 - size_of_last_decoding) + new_match->decoded.GetSize();
+		size_of_last_decoding = new_match->start + new_match->size + new_match->semicolon;
+	}
+
+	Match *create_match(idx_t start, idx_t size, bool semi) {
+		auto input_data = input.GetData();
+		if (input_data[start] == '#') {
+			if (size < 2 || (!isdigit(input_data[start + 1]))) { // true for matches: &, &#, &#<char>..
+				if ((input_data[start + 1] == 'x' || input_data[start + 1] == 'X') && size > 2) {
+					return new HexMatch<HexValue>(input_data, bind_data, start, size, semi);
+				}
+				auto match = new Match(start, size, semi);
+				match->decoded = string_t(input_data + start - 1, size + semi + 1); // return the match
+				return match;
+			}
+			return new HexMatch<UnicodeValue>(input_data, bind_data, start, size, semi);
+		}
+		return new CharRefMatch(input_data, bind_data, start, size, semi);
+	}
+
+public:
+	Match *head;
+	string_t &input;
+	idx_t total_chars_in_list;
+	UnescapeBindData &bind_data;
+
+private:
+	// this is used to calculate the total_chars_in_list
+	idx_t size_of_last_decoding = 0;
+};
+
+static string_t EscapeInputStr(string_t &input_str, bool input_quote, Vector &result) {
+	const auto input_data = input_str.GetData();
+	const auto input_size = input_str.GetSize();
+
+	auto calc_sz = [&]() {
+		idx_t sz = 0;
+		for (idx_t i = 0; i < input_size; i++) {
+			char ch = input_data[i];
+			if (ch == '&') {
+				sz += 5;
```

Comment:
Can we either use `strlen` here instead of `5` or make a constexpr like `constexpr idx_t AMPERSAND_SIZE = 5;` that we can reuse here and in the loop below?
The use of magic numbers just makes me a little uneasy